src/common/src/array/map_array.rs

```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    impl Ord for MapRef<'_> {
+        fn cmp(&self, _other: &Self) -> Ordering {
+            unreachable!("map is not comparable. Such usage should be banned in frontend.")
+        }
+    }
+
+    impl PartialOrd for MapRef<'_> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+}
+
+impl Debug for MapValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.as_scalar_ref().fmt(f)
+    }
+}
+
+impl Display for MapValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.as_scalar_ref().write(f)
+    }
+}
+
+impl MapValue {
+    pub fn from_list_entries(list: ListValue) -> Self {
+        if cfg!(debug_assertions) {
+            // validates list type is valid
+            _ = MapType::from_list_entries(list.data_type());
+        }
+        // TODO: validate the values is valid
+        MapValue(list)
+    }
+
+    pub fn from_kv(keys: ListValue, values: ListValue) -> Self {
+        if cfg!(debug_assertions) {
+            assert_eq!(
+                keys.len(),
+                values.len(),
+                "keys: {keys:?}, values: {values:?}"
+            );
+            let unique_keys = keys.iter().unique().collect_vec();
+            assert!(
+                unique_keys.len() == keys.len(),
+                "non unique keys in map: {keys:?}"
+            );
+            assert!(!unique_keys.contains(&None), "null key in map: {keys:?}");
+        }
+
+        let len = keys.len();
+        let key_type = keys.data_type();
+        let value_type = values.data_type();
+        let struct_array = StructArray::new(
+            MapType::struct_type_for_map(key_type, value_type),
+            vec![keys.into_array().into_ref(), values.into_array().into_ref()],
+            Bitmap::ones(len),
+        );
+        MapValue(ListValue::new(struct_array.into()))
+    }
+}
+
+/// A map is just a slice of the underlying struct array.
+///
+/// XXX: perhaps we can make it `MapRef<'a, 'b>(ListRef<'a>, ListRef<'b>);`.
+/// Then we can build a map ref from 2 list refs without copying the data.
+/// Currently it's impossible.
+#[derive(Copy, Clone, Eq)]
+pub struct MapRef<'a>(pub(crate) ListRef<'a>);
+
+impl<'a> MapRef<'a> {
+    /// Iterates over the elements of the map.
+    pub fn iter(
+        self,
+    ) -> impl DoubleEndedIterator + ExactSizeIterator<Item = (ScalarRefImpl<'a>, DatumRef<'a>)> + 'a
+    {
+        self.0.iter().map(|list_elem| {
+            let list_elem = list_elem.expect("the list element in map should not be null");
+            let struct_ = list_elem.into_struct();
+            let (k, v) = struct_
+                .iter_fields_ref()
+                .next_tuple()
+                .expect("the struct in map should have exactly 2 fields");
+            (k.expect("map key should not be null"), v)
```

Comment:
Where will we enforce the non-nullability of the key?

Reply:
I also felt troubled that it seems we can enforce in too many places (both `MapValue`, `MapRef` and `MapArray`). Do you have any ideas on how to enforce in one place?

Reply:
We can define the types in a separated `mod`. Then even `mod map_array` cannot construct them directly, and every construction need to be explicit. 😄